### PR TITLE
Change log header/payload options

### LINKED ISSUE #336: Change Log Headers Options
With more headers and payloads being added via the "wreck" event, the `includeRequestPayload` type options are becoming a little cumbersome to work with. We need a more maintainable way to control what is sent out from good.

### PR DESCRIPTION
**Pull request by Georgette**
This changes the options `requestHeaders`, `requestPayload`, and `responsePayload` to accept arrays instead of booleans. It also introduces a new option, `responseHeaders`. These arrays must consist of 0 or more strings of `response`, or `wreck`. 

For example, specifying:

```
{
  requestHeaders: ['response', 'wreck'],
  responseHeaders: ['response'],
  requestPayload: ['response'],
  responsePayload: ['response', 'wreck']
}
```

Will result in request headers being available for both `response` and `wreck` events, response headers being available for only `response` events, request payloads being available for `response` events, and response payloads being available for both `response` and `wreck` events.

At this time, it is not possible to make request payloads available to `wreck` events. I was unable to identify how to capture the payload from the `response` event emitted by the Wreck singleton.

In addition, the Good `filter` option also applies to `wreck` response payloads. 

Test coverage remains at 100% with these changes.

Fixes #336

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -22,9 +22,10 @@ set `options` to an object with the following optional settings:
 
 - `[httpAgents]` - the list of `httpAgents` to report socket information about. Can be a single `http.Agent` or an array of agents objects. Defaults to `Http.globalAgent`.
 - `[httpsAgents]` - the list of `httpsAgents` to report socket information about. Can be a single `https.Agent` or an array of agents. Defaults to `Https.globalAgent`.
-- `[requestHeaders]` - determines if all request headers will be available to `reporter` objects. Defaults to _false_
-- `[requestPayload]` - determines if the request payload will be available to `reporter` objects. Defaults to _false_
-- `[responsePayload]` - determines if the response payload will be available to `reporter` objects. Defaults to _false_
+- `[requestHeaders]` - determines if all request headers will be available to `reporter` objects. Defaults to `[]`. Valid array values are `wreck` and/or `response`. The request headers will be included in the events specified within the array.
```
Comment (arb) [inline README.md]: Can we shorten these to `req` and `res` instead of the whole word?

File: lib/monitor.js [hunk 2]
```diff
@@ -239,7 +241,22 @@ internals.Monitor.prototype._opsHandler = function (results) {
 
 internals.Monitor.prototype._wreckHandler = function (error, request, response, start, uri) {
 
-    this._dataStream.push(Utils.GreatWreck(error, request, response, start, uri));
+    var self = this;
```
Comment (arb) [inline lib/monitor.js]: I don't really like this change. If you notice there isn't any `JSON.parse` anywhere else in the code as that is left to the individual reporters to deal with. Also the `try, catch` operation in this function will prevent this code from being optimized by v8.

File: lib/utils.js [hunk 3]
```diff
@@ -141,22 +142,26 @@ exports.GreatResponse = function (request, options, filterRules) {
 
     this.log = request.getLog();
 
-    if (options.requestHeaders) {
+    if (~options.requestHeaders.indexOf('response')) {
```
Comment (arb) [inline lib/utils.js]: For the sake of readability, just do the `-1` check; the next person might not understand what the code is doing.

File: test/monitor.js [hunk 4]
```diff
@@ -492,6 +494,70 @@ describe('good', function () {
             });
         });
 
+        // it('accepts "response" event within "requestHeaders", "responseHeaders", "requestPayload", and "responsePayload" as configuration arrays', function (done) {
```
Comment (arb) [inline test/monitor.js]: Remove commented out code.

File: lib/monitor.js [hunk 5]
```diff
@@ -239,7 +241,22 @@ internals.Monitor.prototype._opsHandler = function (results) {
...
+    var options = this._includeOptions;
+
+    if (~options.responsePayload.indexOf('wreck')) {
+        Wreck.read(response, {}, function (error, payload) {
```
Comment (arb) [inline lib/monitor.js]: You can pass options into `Wreck.read` to have it handle the JSON parsing.

https://github.com/hapijs/wreck/blob/master/lib/index.js#L252

File: lib/utils.js [hunk 6]
```diff
@@ -241,15 +246,29 @@ exports.GreatWreck = function (error, request, response, start, uri) {
...
     };
 
+    if (~options.responseHeaders.indexOf('wreck')) {
+        this.response.headers = response.headers;
+    }
+
+    if (payload) {
+        this.response.payload = payload;
+        if (Object.keys(filterRules).length) {
+
```
Comment (arb) [inline lib/utils.js]: Don't need a new line here.

**Comment by arb** [comment]
If you can't get the request payloads available to wreck events, please update the code and docs accordingly.

File: lib/schema.js [hunk 7]
```diff
@@ -14,9 +14,10 @@ exports.assert = function (type, options) {
 internals.monitorOptions = Joi.object().keys({
     httpAgents: Joi.array(),
     httpsAgents: Joi.array(),
-    requestHeaders: Joi.boolean(),
-    requestPayload: Joi.boolean(),
-    responsePayload: Joi.boolean(),
+    requestHeaders: Joi.array().items(Joi.string().valid('response', 'wreck')),
+    responseHeaders: Joi.array().items(Joi.string().valid('response', 'wreck')),
+    requestPayload: Joi.array().items(Joi.string().valid('response', 'wreck')),
```
Comment (arb) [inline lib/schema.js]: If it doesn't work for "wreck", then it should not be in the list of `valid`s

**Comment by Georgette** [comment]
Thanks for looking through this, I'll make the suggested changes today. Question for you, is it currently feasible to capture the payload from the response event emitted by the Wreck singleton, and perhaps I'm overlooking it?

**Comment by geek** [comment]
@Georgette in the wreck event the following params are passed to the handler `err, req, res`.  The `req` and `res` should both contain the request payload and response body.

**Comment by Georgette** [comment]
For the wreck response event, I'm getting a `ClientHttp` object for request, and `IncomingMessage` for response. I'm still not sure where to find the request payload in either of these. Further hand-holding appreciated!

File: lib/utils.js (hunk 3 above)
Comment (Georgette) [inline lib/utils.js]: Corrected

File: test/monitor.js (hunk 4 above)
Comment (Georgette) [inline test/monitor.js]: corrected

File: lib/monitor.js (hunk 5 above)
Comment (Georgette) [inline lib/monitor.js]: corrected

**Comment by arb** [comment]
@Georgette, thanks for the PR. I am going to close this as I'm not ready to implement another breaking change to good and I'm not sure about the stuff with `wreck` as that seems to be in flux at present (see https://github.com/hapijs/good/pull/386).

**Pull request closed without merging**